### PR TITLE
Add load shedder integration tests to V4 version

### PR DESCRIPTION
**Pull request by tonidero**
### Description
We recently created the new load shedder server to take charge in case the main servers fail. This mechanism is a failsafe to make sure we don't lose any data and most customers aren't affected by this downtime. In order to ensure that these emergency system works correctly when we need it, we plan to run periodic tests that hit the load shedder and make sure the SDKs keep working as expected.

This PR adds integration tests for the load shedder for SDK version 4. Most of the changes are based off #899 and #911 

This is pointing to branch `v4-load-shedder-integration-tests-do-not-remove`. This branch will remain open with the changes needed to run these integration tests against the load shedder. And we will run them periodically from that branch, but they won't get deployed in an SDK version.

#### TODO (in other PRs)
- [x] Add automation #934 
- [x] Add daily schedule to run these tests #934 
- [ ] Add slack notifications

### CODE REVIEW

File: library.gradle [hunk 1]
```diff
@@ -41,4 +62,4 @@ dependencies {
     testImplementation 'org.assertj:assertj-core:3.13.2'
 }
 
-apply plugin: 'com.vanniktech.maven.publish'
+//apply plugin: 'com.vanniktech.maven.publish'
```
Comment (tonidero) [inline library.gradle:65]: I wasn't able to get the maven plugin to work with the new flavors... I removed it for now

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 2]
```diff
@@ -1844,7 +1844,8 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
         @JvmStatic
         @JvmSynthetic
         internal fun configure(
-            configuration: PurchasesConfiguration
+            configuration: PurchasesConfiguration,
+            overrideBillingAbstract: BillingAbstract? = null
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:1848]: Note how this is actually part of the main flavor. However this method is internal, so it's not exposed to the public. So I think it's ok

File: library.gradle [hunk 3]
```diff
@@ -6,6 +6,27 @@ android {
...
 
+    flavorDimensions "extraAPIs"
+    productFlavors {
+        defaults {
+            dimension "extraAPIs"
+        }
+        integrationTest {
+            dimension "extraAPIs"
+            minSdkVersion 21 // Needed because mockk-android requires API 21
+            compileSdkVersion 31
```
Comment (tonidero) [inline library.gradle:17]: Had to increase this version as well due to incompatibilities with some of the testing libraries we use...

File: build.gradle [hunk 4]
```diff
@@ -1,5 +1,5 @@
 buildscript {
-    ext.kotlinVersion = "1.4.10"
+    ext.kotlinVersion = "1.6.21"
```
Comment (tonidero) [inline build.gradle:2]: If this was to be merged, this should be much more thoroughly tested, but it isn't for now. This is needed by some of the testing dependencies

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        val purchaseDetails = PurchaseDetailsFactory.createPurchaseDetails()
+        mockBillingAbstract.mockQuerySkuDetails(querySkuDetailsSubsReturn = listOf(productDetails))
+
+        onActivityReady { activity ->
+            Purchases.sharedInstance.purchaseProductWith(
+                activity = activity,
+                skuDetails = productDetails.skuDetails,
+                onError = { error, _ -> fail("Purchase should be successful. Error: ${error.message}") },
+                onSuccess = { _, customerInfo ->
+                    assertThat(customerInfo.allPurchaseDatesByProduct.containsKey(productDetails.sku)).isTrue
```
Comment (aboedo) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt:149]: should we check that the entitlement is there too? I mean, I get that that this gets indirectly calculated from the entitlement info, but still

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    @Test
+    fun customerInfoCanBeFetched() {
+        val lock = CountDownLatch(1)
+
+        onActivityReady {
+            Purchases.sharedInstance.getPurchaserInfoWith(
+                onError = { fail("should be success. Error: ${it.message}") },
+                onSuccess = { lock.countDown() }
```
Comment (aboedo) [inline purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt:105]: wait, how is this working? This should 5xx, right?

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt (hunk 5 above)
Comment (tonidero) [reply purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt:149]: I'm waiting for the fix for that to be merged into the load shedder. Our test app uses the new product id format, which is causing entitlements not to be granted. Once that's fixed, I need to revisit integration tests for all 3 versions and add that test.

File: purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt (hunk 6 above)
Comment (tonidero) [reply purchases/src/androidTestIntegrationTest/kotlin/com/revenuecat/purchases/PurchasesIntegrationTest.kt]: That request is not being redirected to the load shedder, that's why it works. In any case, that test isn't needed for the load shedder so I removed it.

**Comment by tonidero** [comment]
> reminder to add a branch protection rule so we don't accidentally kill it

Done!

**Review by aboedo — APPROVED** [review]

**Pull request merged**